src/librustc_typeck/check/coercion.rs

```diff
@@ -1282,7 +1290,16 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
                     .filter(|e| fcx.is_assign_to_bool(e, self.expected_ty()))
                     .is_some();
 
-                err.emit_unless(assign_to_bool);
+                if unsized_return {
+                    fcx.tcx.sess.delay_span_bug(
+                        cause.span,
+                        &format!(
+                            "elided E0308 in favor of more detailed E0277 or E0746: {:?}",
+                            cause.code
+                        ),
+                    );
+                }
+                err.emit_unless(assign_to_bool || unsized_return);
```

Comment:
This will likely be the most controversial change: when we have an error that would emit both an "unsized return type" and a "mismatched types" error, it will only emit the former. In a sample case where a function `fn foo() -> dyn T` with three returns, this change emits a single error instead of four (unsized return + 3 type mismatches).

Reply:
Seems no worse to me than `assign_to_bool`.